### PR TITLE
IEI-118580 correct encoding with value length < 64 Kib.  Bring fix from master into generic-config branch.

### PR DESCRIPTION
**Pull request by alan-farquhar**
There is an issue with the generic-config branch handling large tags.

### CODE REVIEW

File: dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java [hunk 1]
```diff
@@ -170,6 +170,9 @@ public void writeHeader(int tag, VR vr, int len) throws IOException {
         ByteUtils.tagToBytes(tag, b, 0, bigEndian);
         int headerLen;
         if (!TagUtils.isItem(tag) && explicitVR) {
+            if ((len & 0xffff0000) != 0 && vr.headerLength() == 8) {
+                vr = VR.UN;
+            }
```
Comment (andrew-cowan) [inline dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java:175]: Add a comment in to clarify what you're doing here:  DCM4CHE has a lot of bit masks and binary math :)

**Comment by hczedik** [comment]
This is a cherry-pick of https://github.com/dcm4che/dcm4che/commit/14eff3a0d46d7632590148087fb718b1f8d393a5 from the master branch. Would be nice to actually do it as a cherry-pick, and by that keep Gunter as the actual Author of the commit.

**Comment by alan-farquhar** [comment]
> This is a cherry-pick of [14eff3a](https://github.com/dcm4che/dcm4che/commit/14eff3a0d46d7632590148087fb718b1f8d393a5) from the master branch. Would be nice to actually do it as a cherry-pick, and by that keep Gunter as the actual Author of the commit.
> 
> Fixes #421 which is already solved on the master branch.

Okay... i will try to figure out how to do it as a cherry-pick.

**Review by craigtmoore — APPROVED** [review]

File: dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java (hunk 1 above)
Comment (alan-farquhar) [reply dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java]: Ok Andrew i've cherry picked Gunther's change from master and added a comment to it.

**Comment by alan-farquhar** [comment]
> > This is a cherry-pick of [14eff3a](https://github.com/dcm4che/dcm4che/commit/14eff3a0d46d7632590148087fb718b1f8d393a5) from the master branch. Would be nice to actually do it as a cherry-pick, and by that keep Gunter as the actual Author of the commit.
> > Fixes #421 which is already solved on the master branch.
> 
> Okay... i will try to figure out how to do it as a cherry-pick.

@hczedik I've cherry picked the change from master now and added a comment as requested by Andrew.

File: dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java [hunk 2]
```diff
@@ -170,6 +170,10 @@ public void writeHeader(int tag, VR vr, int len) throws IOException {
         ByteUtils.tagToBytes(tag, b, 0, bigEndian);
         int headerLen;
         if (!TagUtils.isItem(tag) && explicitVR) {
+            // Check to see if the size of the value to see if it is greater than 64Kib (8th bit is needed).
+            // If so, change the VR tag to unsigned before encoding the value.
```
Comment (hczedik) [inline dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java:174]: This comment explains what the code is doing (which should already be obvious from the code), but not the "why".
The issue description from Gunter explains it quite well:

> Encode Attributes whose known Value Representation is none of OB, OD, OF, OW, SQ, UC, UR or UT and whose value length exceeds 65534 (2^16-2) - and therefore cannot be encoded as a
> 16-bit unsigned integer in the Value Length Field defined for the known Value Representation - with Value Representation UN, according DICOM CP 1066 - ftp://medical.nema.org/medical/dicom/final/cp1066_ft.pdf

If you want to add a comment, that one might make more sense,

**Review by hczedik — APPROVED** [review]

File: dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java (hunk 2 above)
Comment (hczedik) [reply dcm4che-core/src/main/java/org/dcm4che3/io/DicomOutputStream.java:174]: To stay in sync with master, such changes (adding comments) should also be committed to master (via a PR)!

**Review by andrew-cowan — APPROVED** [review]

**Pull request merged**